### ISSUE TITLE
Angular Universal: HTMLElement is not defined

### ISSUE THREAD

**Issue by juristr** [angular/angular#39950]
# 🐞 bug report

### Affected Package

 The issue is caused by package @angular/elements

### Description

Using Angular Elements (specifically `createCustomElement`) breaks SSR.

## 🔬 Minimal Reproduction

Here's a repository that reproduces the issue: https://github.com/juristr/repro-angular-elements-with-ssr-issue

It's a newly generated Angular CLI app + with SSR using `@nguniversal/hapi-engine`.

This commit works [96f5ed9fc5f40ab61be09854200ff404090698e1](https://github.com/juristr/repro-angular-elements-with-ssr-issue/tree/96f5ed9fc5f40ab61be09854200ff404090698e1), but once I add `@angular/elements` it breaks.

## 🔥 Exception or Error
<pre><code>
/Users/.../ngssr/dist/ngssr/server/main.js:55010
class NgElement extends HTMLElement {
                        ^

ReferenceError: HTMLElement is not defined
    at Module.Dti6 (/Users/.../ngssr/dist/ngssr/server/main.js:55010:25)
    at __webpack_require__ (/Users/.../ngssr/dist/ngssr/server/main.js:20:30)
    at Module.ZAI4 (/Users/.../ngssr/dist/ngssr/server/main.js:78554:75)
    at __webpack_require__ (/Users/.../ngssr/dist/ngssr/server/main.js:20:30)
    at Module.24aS (/Users/.../ngssr/dist/ngssr/server/main.js:38572:69)
    at __webpack_require__ (/Users/.../ngssr/dist/ngssr/server/main.js:20:30)
    at Module.K011 (/Users/.../ngssr/dist/ngssr/server/main.js:61050:80)
    at __webpack_require__ (/Users/.../ngssr/dist/ngssr/server/main.js:20:30)
    at Module.uj+Y (/Users/.../ngssr/dist/ngssr/server/main.js:155539:74)
    at __webpack_require__ (/Users/.../ngssr/dist/ngssr/server/main.js:20:30)

A server error has occurred.
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 11.0.2
Node: 12.14.0
OS: darwin x64

Angular: 11.0.2
... animations, cli, common, compiler, compiler-cli, core
... elements, forms, platform-browser, platform-browser-dynamic
... platform-server, router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1100.2
@angular-devkit/build-angular   0.1100.2
@angular-devkit/core            11.0.2
@angular-devkit/schematics      11.0.2
@nguniversal/builders           11.0.0
@nguniversal/hapi-engine        11.0.0
@schematics/angular             11.0.2
@schematics/update              0.1100.2
rxjs                            6.6.3
typescript                      4.0.5
</code></pre>

**Anything else relevant?**

Related:

- https://github.com/angularathens/2nd-workshop-angular-contrib/issues/10#issuecomment-735907510
- https://github.com/angular/angular/issues/24551
- https://github.com/angular/angular/pull/37799

**Comment by gkalpak** [comment]
A little update:

Since #39994 is a breaking change, we are investigating other (non-breaking) solutions, such as assigning `domino.impl` onto `global` sooner. For reference, this assignment currently happens in the [setDomTypes()][1] function, which is [called in a `PLATFORM_INITIALIZER`][2].

In the meantime, a simple work-around is to manually load `domino` sooner:

1. Create a `src/app/load-domino.server.ts` file with the following content:
    ```ts
    Object.assign(global, require('domino').impl);
    ``` 

2. At the top of `src/app/app.server.module.ts` (before any other imports) add an import to the previous file:
    ```ts
    // This must be the first import in this file, so that it is loaded before any (transient)
    // imports to `@angular/elements`, which rely on DOM built-ins being available.
    import './load-domino.server';
    ```

[1]: https://github.com/angular/angular/blob/0fc8466f1be392917e0c7e3448b2144d52286b56/packages/platform-server/src/domino_adapter.ts#L17-L21
[2]: https://github.com/angular/angular/blob/0fc8466f1be392917e0c7e3448b2144d52286b56/packages/platform-server/src/server.ts#L33

**Comment by gkalpak** [comment]
For reference, I have created angular/angular#40559 and angular/angular-cli#19869 to address this issue in a generic way.

**Comment by alan-agius4** [comment]
@gkalpak, I think this can be closed now.

**Comment by gkalpak** [comment]
Argh!!! I put the wrong reference in https://github.com/angular/angular-cli/pull/19869/commits/38b42c29fb27d16ce4988a8fa65ebf5bee07b5e2:

```diff
-Fixes angular/angular#40559
+Fixes angular/angular#39950
```

😇
